### ISSUE TITLE
Issue Report: Discrepancy in acos Function Results in PyTorch Compared to Other Frameworks

### ISSUE THREAD

**Issue by LilyDong0127** [pytorch/pytorch#137860]
### 🐛 Describe the bug

Issue Summary:

During the evaluation of the acos function across different deep learning frameworks, we observed that PyTorch produced significantly different results compared to TensorFlow, Keras, JAX, and Chainer for certain input values. While all other libraries provided consistent outputs, the results from PyTorch exhibited a noticeable discrepancy.
```
import torch
import tensorflow as tf
import jax.numpy as jnp
import numpy as np
from keras.layers import Lambda

# Check if numpy's acos function is accessible
try:
    print(f"Numpy acos function: {np.acos}")
except AttributeError as e:
    print("Numpy acos function is not available:", e)

# Input data (as provided)
input_data = np.array([
    [
        [
            0.5834500789642334,
            0.057789839804172516
        ],
        [
            0.13608910143375397,
            0.8511932492256165
        ],
        [
            -0.8579278588294983,
            -0.8257414102554321
        ],
        [
            -0.9595631957054138,
            0.665239691734314
        ],
        [
            0.5563135147094727,
            0.7400242686271667
        ],
        [
            0.9572367072105408,
            0.5983171463012695
        ],
        [
            -0.07704125344753265,
            0.5610583424568176
        ],
        [
            -0.7634511590003967,
            0.2798420190811157
        ],
        [
            -0.7132934331893921,
            0.8893378376960754
        ]
    ]
])

# PyTorch acos operation
def torch_acos(x):
    return torch.acos(torch.tensor(x, dtype=torch.float32))

# TensorFlow acos operation
def tf_acos(x):
    return tf.acos(tf.convert_to_tensor(x, dtype=tf.float32))

# Keras acos operation
def keras_acos(x):
    return Lambda(lambda x: tf.math.acos(x))(tf.convert_to_tensor(x, dtype=tf.float32))

# JAX acos operation
def jax_acos(x):
    return jnp.arccos(jnp.array(x))

# Chainer acos operation
def chainer_acos(x):
    return np.arccos(x)

# Calculate results
pytorch_result = torch_acos(input_data)
tensorflow_result = tf_acos(input_data)
keras_result = keras_acos(input_data).numpy()  # Convert Keras result to numpy
jax_result = jax_acos(input_data)
chainer_result = chainer_acos(input_data)

# Print results
print(f"PyTorch acos result: {pytorch_result.detach().numpy()}")  # Detach to convert to numpy
print(f"TensorFlow acos result: {tensorflow_result.numpy()}")
print(f"Keras acos result: {keras_result}")
print(f"JAX acos result: {jax_result}")
print(f"Chainer acos result: {chainer_result}")
```
results:
```
PyTorch acos result: [[[0.947826   1.5129743 ]
  [1.4342836  0.55254173]
  [2.602019   2.5423117 ]
  [2.8562427  0.8429815 ]
  [0.9808535  0.7376899 ]
  [0.29350162 0.9293971 ]
  [1.647914   0.9751325 ]
  [2.4394362  1.2871668 ]
  [2.3649826  0.47490135]]]
TensorFlow acos result: [[[0.9478259  1.5129744 ]
  [1.4342835  0.5525418 ]
  [2.6020193  2.5423117 ]
  [2.856243   0.8429814 ]
  [0.9808533  0.73769   ]
  [0.2935016  0.929397  ]
  [1.647914   0.9751324 ]
  [2.4394362  1.2871668 ]
  [2.3649824  0.47490135]]]
Keras acos result: [[[0.9478259  1.5129744 ]
  [1.4342835  0.5525418 ]
  [2.6020193  2.5423117 ]
  [2.856243   0.8429814 ]
  [0.9808533  0.73769   ]
  [0.2935016  0.929397  ]
  [1.647914   0.9751324 ]
  [2.4394362  1.2871668 ]
  [2.3649824  0.47490135]]]
JAX acos result: [[[0.94782597 1.5129744 ]
  [1.4342836  0.55254173]
  [2.602019   2.5423117 ]
  [2.8562427  0.8429815 ]
  [0.9808535  0.73768985]
  [0.29350162 0.92939717]
  [1.6479139  0.9751326 ]
  [2.4394362  1.2871667 ]
  [2.3649824  0.47490135]]]
Chainer acos result: [[[0.947826   1.51297427]
  [1.43428362 0.55254171]
  [2.60201911 2.54231171]
  [2.85624268 0.84298152]
  [0.9808535  0.73768989]
  [0.29350162 0.92939713]
  [1.647914   0.97513254]
  [2.43943616 1.28716678]
  [2.36498246 0.47490134]]]
```
Observation: Upon analyzing the results, it is clear that PyTorch produces results that differ from the outputs of the other frameworks. In particular:

The result for -0.8579278588294983 shows significant differences across frameworks. For instance, while PyTorch's result is 2.602019, TensorFlow and Chainer yield 2.6020193 and 2.60201911, respectively.

### Versions

sys.platform: win32
Python version: 3.9.19 | packaged by conda-forge | (main, Mar 20 2024, 12:38:46) [MSC v.1929 64 bit (AMD64)]
Locale: ('zh_CN', 'cp936')
PyTorch version: 2.2.1+cpu
CUDA available: False
Installed packages: absl-py==2.1.0
astunparse==1.6.3
attrs==23.2.0
black==24.3.0
blinker==1.7.0
build==1.1.1
CacheControl==0.14.0
certifi==2024.2.2
chainer==7.8.1
charset-normalizer==3.3.2
cleo==2.1.0
click==8.1.7
cloudpickle==3.0.0
colorama==0.4.6
coverage==7.4.4
crashtest==0.4.1
dash==2.16.1
dash-core-components==2.0.0
dash-html-components==2.0.0
dash-table==5.0.0
decorator==5.1.1
distlib==0.3.8
dm-tree==0.1.8
dulwich==0.21.7
exceptiongroup==1.2.0
fastjsonschema==2.19.1
filelock==3.13.1
Flask==3.0.2
flatbuffers==24.3.7
fsspec==2024.3.0
gast==0.5.4
google-pasta==0.2.0
graphviz==0.20.1
grpcio==1.62.1
h5py==3.10.0
hypofuzz==24.2.3
hypothesis==6.99.9
hypothesmith==0.3.3
idna==3.6
importlib_metadata==7.0.2
iniconfig==2.0.0
installer==0.7.0
itsdangerous==2.1.2
jaraco.classes==3.3.1
jax==0.4.26
jaxlib==0.4.26
Jinja2==3.1.3
keras==3.2.0
keyring==24.3.1
lark==1.1.9
libclang==16.0.6
libcst==1.2.0
Markdown==3.6
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
ml-dtypes==0.3.1
more-itertools==10.2.0
mpmath==1.3.0
msgpack==1.0.8
mypy-extensions==1.0.0
namex==0.0.7
nest-asyncio==1.6.0
networkx==3.2.1
numpy==1.26.4
opt-einsum==3.3.0
optree==0.11.0
packaging==24.0
pathspec==0.12.1
pexpect==4.9.0
pkginfo==1.10.0
platformdirs==4.2.0
plotly==5.20.0
pluggy==1.4.0
poetry==1.8.2
poetry-core==1.9.0
poetry-plugin-export==1.7.1
protobuf==4.25.3
psutil==5.9.8
ptyprocess==0.7.0
Pygments==2.17.2
pyproject_hooks==1.0.0
pytest==8.1.1
python-dateutil==2.9.0.post0
pytz==2024.1
pywin32-ctypes==0.2.2
PyYAML==6.0.1
rapidfuzz==3.7.0
requests==2.31.0
requests-toolbelt==1.0.0
retrying==1.3.4
rich==13.7.1
scipy==1.13.1
shellingham==1.5.4
six==1.16.0
sortedcontainers==2.4.0
sympy==1.12
tenacity==8.2.3
tensorboard==2.16.2
tensorboard-data-server==0.7.2
tensorflow==2.16.1
tensorflow-intel==2.16.1
tensorflow-io-gcs-filesystem==0.31.0
tensorflow-probability==0.24.0
termcolor==2.4.0
tf_keras==2.16.0
Theano==1.0.5
tomli==2.0.1
tomlkit==0.12.4
torch==2.2.1
torchaudio==2.2.1
torchvision==0.17.1
trove-classifiers==2024.3.3
typing-inspect==0.9.0
typing_extensions==4.11.0
tzdata==2024.1
urllib3==2.2.1
virtualenv==20.25.1
Werkzeug==3.0.1
wrapt==1.16.0
zipp==3.18.1

**Comment by LilyDong0127** [comment]
pytorch符合预期

**Comment by albanD** [comment]
I'm afraid all of these results are within numerical accuracy from the floating point standard. I'm sure that if you check the results on GPU, you will see similar differences even within the same framework.
The difference in this case is most likely due to some framework using vectorization or acceleration libraries in a different way.

> while PyTorch's result is 2.602019, TensorFlow and Chainer yield 2.6020193 and 2.60201911,

Note that the discrepency between this kind of numbers might just be due to printing being done differently haha

If you want more scary case of that:
```
import torch

torch.manual_seed(0)
input_data = torch.randn(100)

torch.set_printoptions(precision=10, profile="full")
print(input_data.cos()[:5])
input_data[3] = 1e10
print(input_data.cos()[:5])
```
Prints this for me locally (yes this is cos, not acos)
```
$ python foo.py 
tensor([0.4304186106, 0.4063319862, 0.9687690735, 0.9073420167, 0.6609514952])
tensor([0.4304186404, 0.4063320160, 0.9687690735, 0.8731196523, 0.6609514952])

```

So the 3rd value is changed as expected. But the other values also change due to how the vectorized implementation of cos is done within the acceleration library.
You should generally not rely on equality and only floating-point level equality https://pytorch.org/docs/stable/notes/numerical_accuracy.html
